File: Java/circular.java
import java.util.Scanner;

class Node {
    int data;
    Node next, prev;
}

public class circular {
    static Node head = null;
    static Scanner sc = new Scanner(System.in);

    static void begininsert() {
        Node ptr = new Node();
        ptr.data = sc.nextInt();
        if (head == null) {
            ptr.next = ptr;
            ptr.prev = ptr;
            head = ptr;
        } else {
            ptr.next = head;
            ptr.prev = head.prev;
            head.prev.next = ptr;
            head.prev = ptr;
            head = ptr;
        }
    }

    static void lastinsert() {
        Node ptr = new Node();
        ptr.data = sc.nextInt();
        if (head == null) {
            ptr.next = ptr;
            ptr.prev = ptr;
            head = ptr;
        } else {
            ptr.next = head;
            ptr.prev = head.prev;
            head.prev.next = ptr;
            head.prev = ptr;
        }
    }

    static void randominsert() {
        Node ptr = new Node();
        ptr.data = sc.nextInt();
        int loc = sc.nextInt();
        Node temp = head;
        for (int i = 0; i < loc; i++) temp = temp.next;
        ptr.next = temp.next;
        ptr.prev = temp;
        temp.next.prev = ptr;
        temp.next = ptr;
    }

    static void begin_delete() {
        if (head == null) return;
        if (head.next == head) {
            head = null;
        } else {
            Node ptr = head;
            head.prev.next = head.next;
            head.next.prev = head.prev;
            head = head.next;
        }
    }

    static void last_delete() {
        if (head == null) return;
        if (head.next == head) {
            head = null;
        } else {
            Node ptr = head.prev;
            ptr.prev.next = head;
            head.prev = ptr.prev;
        }
    }

    static void random_delete() {
        int loc = sc.nextInt();
        Node temp = head;
        for (int i = 0; i < loc; i++) temp = temp.next;
        Node ptr = temp.next;
        temp.next = ptr.next;
        ptr.next.prev = temp;
    }

    static void search_value() {
        int item = sc.nextInt(), pos = 1, found = 0;
        if (head == null) return;
        Node ptr = head;
        do {
            if (ptr.data == item) found = pos;
            ptr = ptr.next;
            pos++;
        } while (ptr != head);
        if (found != 0) System.out.print(found);
        else System.out.print("no");
    }

    static void display() {
        if (head == null) return;
        Node ptr = head;
        do {
            System.out.print(ptr.data + " ");
            ptr = ptr.next;
        } while (ptr != head);
    }

    public static void main(String[] args) {
        int choice;
        while (true) {
            choice = sc.nextInt();
            if (choice == 1) begininsert();
            else if (choice == 2) lastinsert();
            else if (choice == 3) randominsert();
            else if (choice == 4) begin_delete();
            else if (choice == 5) last_delete();
            else if (choice == 6) random_delete();
            else if (choice == 7) search_value();
            else if (choice == 8) display();
            else if (choice == 9) break;
        }
    }
}

File: JavaScript/circular.js
const readline = require("readline").createInterface({
    input: process.stdin,
    output: process.stdout
});

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

let head = null;

function input(q) {
    return new Promise(r => readline.question(q, ans => r(ans)));
}

async function begininsert() {
    let item = parseInt(await input(""));
    let ptr = new Node(item);
    if (head === null) {
        head = ptr;
        ptr.next = ptr;
        ptr.prev = ptr;
    } else {
        let last = head.prev;
        ptr.next = head;
        ptr.prev = last;
        last.next = ptr;
        head.prev = ptr;
        head = ptr;
    }
}

async function lastinsert() {
    let item = parseInt(await input(""));
    let ptr = new Node(item);
    if (head === null) {
        head = ptr;
        ptr.next = ptr;
        ptr.prev = ptr;
    } else {
        let last = head.prev;
        last.next = ptr;
        ptr.prev = last;
        ptr.next = head;
        head.prev = ptr;
    }
}

async function randominsert() {
    let item = parseInt(await input(""));
    let loc = parseInt(await input(""));
    let ptr = new Node(item);
    if (head === null) return;
    let temp = head;
    for (let i = 0; i < loc; i++) {
        temp = temp.next;
        if (temp === head) return;
    }
    let nxt = temp.next;
    ptr.next = nxt;
    ptr.prev = temp;
    temp.next = ptr;
    nxt.prev = ptr;
}

async function begin_delete() {
    if (head === null) return;
    if (head.next === head) {
        head = null;
        return;
    }
    let last = head.prev;
    head = head.next;
    head.prev = last;
    last.next = head;
}

async function last_delete() {
    if (head === null) return;
    if (head.next === head) {
        head = null;
        return;
    }
    let last = head.prev;
    let prev = last.prev;
    prev.next = head;
    head.prev = prev;
}

async function random_delete() {
    let loc = parseInt(await input(""));
    if (head === null) return;
    let temp = head;
    for (let i = 0; i < loc; i++) {
        temp = temp.next;
        if (temp === head) return;
    }
    let ptr = temp.next;
    if (ptr === head) return;
    temp.next = ptr.next
}
